src/finalizer/utils/helios.ts

```
@@ -0,0 +1,927 @@
+
+  const hubPoolStoreInfo = CONTRACT_ADDRESSES[l1ChainId]?.hubPoolStore;
+  if (!hubPoolStoreInfo?.address || !hubPoolStoreInfo.abi) {
+    logger.error({
+      at: `Finalizer#heliosL1toL2Finalizer:getAndFilterL1Events:${l2ChainId}`,
+      message: `HubPoolStore contract address or ABI not found for L1 chain ${l1ChainId}.`,
+    });
+    return null;
+  }
+
+  const hubPoolStoreContract = new ethers.Contract(hubPoolStoreInfo.address, hubPoolStoreInfo.abi as any, l1Provider);
+
+  const l1SearchConfig: EventSearchConfig = {
+    fromBlock: l1SpokePoolClient.eventSearchConfig.fromBlock,
+    toBlock: l1SpokePoolClient.latestBlockSearched,
+    maxBlockLookBack: l1SpokePoolClient.eventSearchConfig.maxBlockLookBack,
+  };
+
+  const storedCallDataFilter = hubPoolStoreContract.filters.StoredCallData();
+
+  try {
+    logger.debug({
+      at: `Finalizer#heliosL1toL2Finalizer:getAndFilterL1Events:${l2ChainId}`,
+      message: `Querying StoredCallData events on L1 (${l1ChainId})`,
+      hubPoolStoreAddress: hubPoolStoreInfo.address,
+      fromBlock: l1SearchConfig.fromBlock,
+      toBlock: l1SearchConfig.toBlock,
+    });
+
+    const rawLogs = await paginatedEventQuery(hubPoolStoreContract, storedCallDataFilter, l1SearchConfig);
+
+    // Explicitly cast logs to the correct type
+    const events: StoredCallDataEvent[] = rawLogs.map((log) => ({
+      ...log,
+      args: log.args as StoredCallDataEventArgs,
+    }));
+
+    const relevantStoredCallDataEvents = events.filter(
+      (event) =>
+        compareAddressesSimple(event.args.target, l2SpokePoolAddress) ||
+        compareAddressesSimple(event.args.target, ethers.constants.AddressZero)
+    );
+
+    logger.debug({
+      at: `Finalizer#heliosL1toL2Finalizer:getAndFilterL1Events:${l2ChainId}`,
+      message: `Found ${relevantStoredCallDataEvents.length} StoredCallData events on L1 (${l1ChainId})`,
+    });
+
+    return relevantStoredCallDataEvents;
+  } catch (error) {
```

why would this fail? I think large try-catch statements are not good to keep around because they can catch unexpected errors also. We should probably just let the error crash the process as I don't see why this should ever fail

---

Agree, this shouldn't be here.

---

I think the try-catch was meant to catch RPC call problems, but not sure `paginatedEventQuery` can ever throw.

---

Actually, yeah it can throw. See https://github.com/across-protocol/relayer/pull/2203#discussion_r2071089377

---

you should just let `paginatedEventQuery` throw like we do in the rest of the codebase. We have specific channels to monitor these alerts and the providers have fallback/quorum logic that should prevent most normal RPC failures.

In general, you should not assume that RPC queries fail and if they do-- your code is too high level to handle it. We should let the provider class do its job